### PR TITLE
RFC-0005: Artifact `Revision` format and introduction of `Digest`

### PR DESCRIPTION
**Pull request by hiddeco**
### TL;DR

This RFC proposes to establish a canonical `Revision` format for an `Artifact`
which points to a specific revision represented as a checksum (e.g. an OCI
manifest digest or Git commit SHA) of a named pointer (e.g. an OCI image tag or
Git tag). In addition, it proposes to include the algorithm name (e.g.
`sha256`) as a prefix to an advertised checksum for an `Artifact` and
further referring to it as a `Digest`, deprecating the `Checksum` field.

### Related to

- https://github.com/fluxcd/source-controller/issues/855
- https://github.com/fluxcd/source-controller/pull/925#issuecomment-1285491112

### CODE REVIEW

File: rfcs/XXXX-artifact-revision-format/README.md [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+main@sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+```
+
+For a `GitRepository`'s `Artifact` pointing towards a specific SHA-1 Git commit
+without a defined branch or tag, the `Revision` field value would become:
+
+```text
+sha1:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+```
+
```
Comment (stefanprodan) [inline rfcs/XXXX-artifact-revision-format/README.md:110]: Can you please explain how this works for Bucket? Is the `checksum` the same with `revision`, do we have the same value for both?

File: rfcs/XXXX-artifact-revision-format/README.md [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+Where `<named pointer>` is the name of e.g. a Git branch or OCI tag,
+`<checksum>` is the exact revision (e.g. a Git commit SHA or OCI manifest
+digest), and `<algo>` is the alias of the algorithm used to calculate the
+checksum (e.g. `sha256`). In case only a named pointer or digest is advertised,
+the `@` is omitted.
+
+For a `GitRepository`'s `Artifact` pointing towards an SHA-256 Git commit on
+branch `main`, the `Revision` field value would become:
```
Comment (stefanprodan) [inline rfcs/XXXX-artifact-revision-format/README.md:98]: Can you add an example for Git tag as well please?

Comment (hiddeco) [reply rfcs/XXXX-artifact-revision-format/README.md:98]: Isn't this described in detail already in the paragraph above this sentence?

File: rfcs/XXXX-artifact-revision-format/README.md [hunk 3]
```diff
@@ -0,0 +1,175 @@
...
+**Last update:** 2022-10-20
+
+## Summary
+
+This RFC proposes to establish a canonical format for an `Artifact` which
+points to a specific checksum (e.g. an OCI manifest digest or Git commit SHA)
+of a named pointer (e.g. an OCI image tag or Git tag). In addition, it proposes
+to include the algorithm name (e.g. `sha256`) as a prefix to any advertised
+checksum in an `Artifact` and further referring to it as a `Digest` opposed to
+a `Checksum`.
```
Comment (pjbgf) [inline rfcs/XXXX-artifact-revision-format/README.md:21]:
```suggestion
checksum in an `Artifact` and further referring to it as a `Digest` as opposed
to a `Checksum`.
```

File: rfcs/XXXX-artifact-revision-format/README.md [hunk 4]
```diff
@@ -0,0 +1,175 @@
...
+- Define a canonical format for an `Artifact`'s `Revision` field which contains
+  a named pointer and a different reference than a checksum.
+
+## Proposal
+
+### Establish an Artifact Revision format
+
+Change the format of the `Revision` field of the `source.toolkit.fluxcd.io`
+Group's `Artifact` type across all `Source` kinds to contain an `@` delimiter
+opposed to `/`, and include the algorithm alias as a prefix to the checksum
```
Comment (pjbgf) [inline rfcs/XXXX-artifact-revision-format/README.md:84]:
```suggestion
as opposed to `/`, and include the algorithm alias as a prefix to the checksum
```

File: rfcs/XXXX-artifact-revision-format/README.md [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+
+- Establish a canonical format to refer to an `Artifact`'s `Revision` field
+  which consists of a named pointer and a specific checksum reference.
+- Allow easier verification of the `Artifact`'s checksum by including an
+  alias for the algorithm.
+- Allow configuration of the algorithm used to calculate the checksum of an
+  `Artifact`.
+- Allow configuration of [BLAKE3][] as an alternative to SHA for calculating
+  checksums. This has promising performance improvements over SHA-256, which
+  could allow for performance improvements in large scale environments.
```
Comment (pjbgf) [inline rfcs/XXXX-artifact-revision-format/README.md:68]: As we are not providing ways for other algorithms to be configured as part of this RFC, we could rather signal the support of them instead...

```suggestion
- Support alternative algorithms for calculating checksums (e.g. [BLAKE3][]).
```

File: rfcs/XXXX-artifact-revision-format/README.md (hunk 2 above)
Comment (stefanprodan) [reply rfcs/XXXX-artifact-revision-format/README.md:98]: Ok

**Review by pjbgf — APPROVED** [review]
LGTM

Great job @hiddeco!

**Comment by hiddeco** [comment]
@pjbgf I will count your approval as sponsorship (albeit [not fully required](https://github.com/fluxcd/flux2/tree/main/rfcs#rfc-process)) and assign `0005` as a reference to this RFC.

**Review by stefanprodan — APPROVED** [review]
LGTM

Thanks @hiddeco 🏅

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 6]
```diff
@@ -0,0 +1,352 @@
...
+
+The current `Artifact` type's `Revision` field format is not "officially"
+standardized (albeit assumed throughout our code bases), and has mostly been
+derived from `GitRepository` which uses `/` as a delimiter between the named
+pointer (a Git branch or tag) and a specific (SHA-1, or theoretical SHA-256)
+revision.
+
+Since the introduction of `OCIRepository` and with the recent changes around
+`HelmChart` objects to allow the consumption of charts from OCI registries.
+This could be seen as outdated or confusing due to the format differing from
```
Comment (makkes) [inline rfcs/0005-artifact-revision-and-digest/README.md:33]:
```suggestion
`HelmChart` objects to allow the consumption of charts from OCI registries,
this could be seen as outdated or confusing due to the format differing from
```

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 7]
```diff
@@ -0,0 +1,352 @@
...
+
+- Establish a canonical format to refer to an `Artifact`'s `Revision` field
+  which consists of a named pointer and a specific checksum reference.
+- Allow easier verification of the `Artifact`'s checksum by including an
+  alias for the algorithm.
+- Allow configuration of the algorithm used to calculate the checksum of an
+  `Artifact`.
+- Allow configuration of [BLAKE3][] as an alternative to SHA for calculating
+  checksums. This has promising performance improvements over SHA-256, which
+  could allow for performance improvements in large scale environments.
```
Comment (makkes) [inline rfcs/0005-artifact-revision-and-digest/README.md:68]: Why do we mix implementation details (i.e. a specific hash function) into this RFC?

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 8]
```diff
@@ -0,0 +1,352 @@
...
+```
+
+For a `Bucket`'s `Artifact` with a revision based on an SHA-256 calculation of
+a list of object keys and their etags, the `Revision` field value would become:
+
+```text
+sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+```
+
+### Change the Artifact Checksum to a Digest
```
Comment (makkes) [inline rfcs/0005-artifact-revision-and-digest/README.md:118]: I'm not sure I understand why we still need a `checksum` field when the `revision` field already contains that checksum.

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+$ kubectl get gitrepository -o jsonpath='{.status.artifact.revision}' <name>
+main@sha256:e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
+```
+
+#### Artifact checksum verification
+
+> As a user of the source-controller, I want to be able to verify the checksum
+> of an Artifact.
+
+For a Source kind with an `Aritfact` the digest consisting of the algorithm
```
Comment (makkes) [inline rfcs/0005-artifact-revision-and-digest/README.md:157]:
```suggestion
For a Source kind with an `Artifact` the digest consisting of the algorithm
```

File: rfcs/0005-artifact-revision-and-digest/README.md (hunk 8 above)
Comment (hiddeco) [reply rfcs/0005-artifact-revision-and-digest/README.md:118]: There is a difference between a revision checksum and the digest. A revision checksum is an _upstream reference_, where as the digest represents the checksum of the advertised file.

Comment (makkes) [reply rfcs/0005-artifact-revision-and-digest/README.md:118]: Ha, missed that. Maybe you can change the examples so that the `revision` and the `checksum` fields are different then where it's appropriate.

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 10]
```diff
@@ -0,0 +1,352 @@
...
+-->
+
+**Creation date:** 2022-10-20
+
+**Last update:** 2022-10-24
+
+## Summary
+
+This RFC proposes to establish a canonical format for an `Artifact` which
+points to a specific checksum (e.g. an OCI manifest digest or Git commit SHA)
```
[the remark this answers is not in the document]
Comment (hiddeco) [reply rfcs/0005-artifact-revision-and-digest/README.md:17]: No, this specifically talks about a "checksum" or a "hash" as an arbitrary revision (which may not be algorithm based) can not be represented as a digest.

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 11]
```diff
@@ -0,0 +1,352 @@
...
+of a named pointer (e.g. an OCI image tag or Git tag). In addition, it proposes
+to include the algorithm name (e.g. `sha256`) as a prefix to any advertised
+checksum in an `Artifact` and further referring to it as a `Digest` as opposed
+to a `Checksum`.
```
[the remark this answers is not in the document]
Comment (hiddeco) [reply rfcs/0005-artifact-revision-and-digest/README.md]: Please see the updating wording of the summary.

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 12]
```diff
@@ -0,0 +1,352 @@
...
+The current `Artifact` type's `Revision` field format is not "officially"
+standardized (albeit assumed throughout our code bases), and has mostly been
+derived from `GitRepository` which uses `/` as a separator between the named
+pointer (a Git branch or tag) and a specific (SHA-1, or theoretical SHA-256)
+revision.
+
+Since the introduction of `OCIRepository` and with the recent changes around
+`HelmChart` objects to allow the consumption of charts from OCI registries,
+this could be seen as outdated or confusing due to the format differing from
+the canonical format used by OCI, which is `<tag>@<algo>:<checksum>` (the
```
Comment (souleb) [inline rfcs/0005-artifact-revision-and-digest/README.md:34]: I am not sure I understand the meanings of `OCI tag` in the document.

The way to  address an `image manifest` in the different OCI specs is  using a `name` and a `reference`.  The `distribution spec` states:
>  \<name\> refers to the namespace of the repository. \<reference\> MUST be either (a) the digest of the manifest or (b) a tag.

We would then have either `<name>:<tag>` or `<name>@sha256:<digest>`.

My understanding is that we want to use the second form, `<name>@sha256:<digest>`.

Comment (hiddeco) [reply rfcs/0005-artifact-revision-and-digest/README.md:34]: You are right, I mixed something up while writing this bit and skimming through the OCI specs. The correct format is `<name>@<digest>`, as a "digest" includes the algorithm the checksum was calculated with (see [here](https://github.com/opencontainers/image-spec/blob/main/descriptor.md#digests)). Which is why I used `@<algo>:<checksum>`.

I'll replace the tag reference(s) with "name".

**Review by souleb — APPROVED** [review]
LGTM. Thanks @hiddeco

**Review by darkowlzz — APPROVED** [review]
LGTM!

File: rfcs/0005-artifact-revision-and-digest/README.md [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+Change the format of the `Revision` field of the `source.toolkit.fluxcd.io`
+Group's `Artifact` type across all `Source` kinds to contain an `@` separator
+opposed to `/`, and include the algorithm alias as a prefix to the checksum
+(creating a "digest").
+
+```text
+[ <named pointer> ] [ [ "@" ] <algo> ":" <checksum> ]
+```
+
+Where `<named pointer>` is the name of e.g. a Git branch or OCI repository
```
Comment (phillebaba) [inline rfcs/0005-artifact-revision-and-digest/README.md:91]: Having the named pointer either be the branch name or OCI repository is a bit weird in my head. These two values represent different things. Its like including the repository URL in the named pointer instead of the Git branch.

**Review by phillebaba — COMMENTED** [review]
Everything looks good I just have a small comment, but I am fine with dropping it if you do not agree with me.

File: rfcs/0005-artifact-revision-and-digest/README.md (hunk 13 above)
Comment (hiddeco) [reply rfcs/0005-artifact-revision-and-digest/README.md:91]: The scope does not explicitly prohibit having a format like e.g. `image-name:tag@sha256:62d3f07430651f515089c402fd006fc0c63382a7fc1087adda3becbb6b4481a3`.

Comment (phillebaba) [reply rfcs/0005-artifact-revision-and-digest/README.md:91]: Ok sounds good.

**Review by phillebaba — APPROVED** [review]
LGTM
